### DISCUSSION TITLE
How to set up a updated boundary condition?

### DISCUSSION THREAD

**Discussion by hhy2022** [idaholab/moose#25442] [Q&A General]
Hi there,

I want to solve a species transport problem when the flow is a loop, which means there is not a traditional left and right boudnary conditions. Instead, the "boundary" conditions should update at each loop. Does there any case that I can learn from or do you have any idea on how to set up such a boundary conditions?

Thank you very much.

**Comment by lindsayad** [comment]
How would you want the conditions to be updated?

**Reply by GiudGiud** [reply]
I think the best way to do a closed loop is to use a mesh that represents the loop. All the cells are connected naturally. 
Then you use a volume region to provide a momentum source, a sort of volumetric pump

**Reply by hhy2022** [reply]
I think there are two possible ways to do so:
1. I have a left and right boundary condition but the boundary conditions should be updated at each time steps. The ith time steps output should be the (i+1)th time steps input. So I need a updated boundary conditions. I am not sure if this is possible in MOOSE setup.
2. I have a naturally connect mesh loop. In this situation, I will not have a left and right boundary conditions. I only have an initail condition. Here I am not sure if this is going to work without boundary conditions in MOOSE.
Besides, when you said a "momentum source", did you mean an instant input? It only work at the beginning of the running and then dispear. Does it mean "initial condition"?

**Reply by lindsayad** [reply]
> I have a left and right boundary condition but the boundary conditions should be updated at each time steps. The ith time steps output should be the (i+1)th time steps input. So I need a updated boundary conditions. I am not sure if this is possible in MOOSE setup.

There are various ways to track older states in MOOSE. For instance there is [slnOld](https://mooseframework.inl.gov/docs/doxygen/moose/classMooseVariableFE.html#a1392f1e407e5b679794cd1d5a9916510) on the variable classes. There are stateful material properties although those would only work with integrated boundary conditions. You can introduce time dependence through [Function](https://mooseframework.inl.gov/syntax/Functions/index.html) objects.

**Reply by hhy2022** [reply]
That makes sense. The only thing is I still don't know the expression of the boundary conditions at this time step since it depends on the former time steps. So the input (boundary condition) of this time steps should be the output of former time steps. In this case, it can be physically forming a loop. In this situation, is there any possible way to do so?

**Reply by GiudGiud** [reply]
To retrieve an old value of a variable, you can use:
- coupledValueOld from the coupleable interface (used by most kernels & bcs)
OR
- slnOld (different API, directly from the variable class)
OR
- the old state arguments, through the FunctorAPI

All of this can work, but I m thinking the big difficulty here is not getting the old state, but matching the two boundaries.
Can you paste an image of your mesh? and highlight the two boundaries you are trying to match

**Reply by hhy2022** [reply]
I am on the progress of designing the meshes. I am considering designing a tube mesh and use the right boundary output as the input of  left boundary at nest time step. This means I will have to update the boundary conditions each time. I am not sure if this will work.
